File: clients/falcondb-python/src/falcondb/connection.py
"""FalconDB native protocol connection."""

from __future__ import annotations

import os
import socket
import ssl
import struct
from typing import Any

from falcondb.errors import AuthenticationError, FalconError, make_error
from falcondb.wire import (
    AUTH_PASSWORD,
    FEATURE_BATCH_INGEST,
    FEATURE_BINARY_PARAMS,
    FEATURE_EPOCH_FENCING,
    FEATURE_PIPELINE,
    FEATURE_TLS,
    MAX_FRAME_SIZE,
    MSG_AUTH_FAIL,
    MSG_AUTH_OK,
    MSG_AUTH_REQUEST,
    MSG_AUTH_RESPONSE,
    MSG_BATCH_REQUEST,
    MSG_BATCH_RESPONSE,
    MSG_CLIENT_HELLO,
    MSG_DISCONNECT,
    MSG_ERROR_RESPONSE,
    MSG_PING,
    MSG_PONG,
    MSG_QUERY_REQUEST,
    MSG_QUERY_RESPONSE,
    MSG_SERVER_HELLO,
    SESSION_AUTOCOMMIT,
    WireReader,
    WireWriter,
    build_frame,
    decode_row,
    encode_row,
)
from falcondb.cursor import ColumnMeta, Cursor


def connect(
    host: str = "localhost",
    port: int = 6543,
    database: str = "falcon",
    user: str = "falcon",
    password: str = "",
    *,
    connect_timeout: float = 10.0,
    ssl_enabled: bool = False,
    ssl_verify: bool = True,
    autocommit: bool = True,
) -> Connection:
    """Create a new connection to a FalconDB server."""
    return Connection(
        host=host,
        port=port,
        database=database,
        user=user,
        password=password,
        connect_timeout=connect_timeout,
        ssl_enabled=ssl_enabled,
        ssl_verify=ssl_verify,
        autocommit=autocommit,
    )


class Connection:
    """A single connection to a FalconDB server using the native protocol."""

    def __init__(
        self,
        host: str = "localhost",
        port: int = 6543,
        database: str = "falcon",
        user: str = "falcon",
        password: str = "",
        connect_timeout: float = 10.0,
        ssl_enabled: bool = False,
        ssl_verify: bool = True,
        autocommit: bool = True,
    ) -> None:
        self._host = host
        self._port = port
        self._database = database
        self._user = user
        self._password = password
        self._autocommit = autocommit
        self._ssl_enabled = ssl_enabled
        self._ssl_verify = ssl_verify

        self._sock: socket.socket | None = None
        self._closed = False
        self._request_id = 0
        self._server_epoch: int = 0
        self._server_node_id: int = 0
        self._negotiated_features: int = 0

        self._connect(connect_timeout)

    # ── Public API ───────────────────────────────────────────────────

    @property
    def autocommit(self) -> bool:
        return self._autocommit

    @autocommit.setter
    def autocommit(self, value: bool) -> None:
        self._autocommit = value

    @property
    def server_epoch(self) -> int:
        return self._server_epoch

    @property
    def server_node_id(self) -> int:
        return self._server_node_id

    @property
    def closed(self) -> bool:
        return self._closed

    def cursor(self) -> Cursor:
        """Create a new cursor bound to this connection."""
        self._check_open()
        return Cursor(self)

    def execute(self, sql: str, params: list[Any] | None = None) -> Cursor:
        """Shorthand: create a cursor, execute, and return it."""
        cur = self.cursor()
        cur.execute(sql, params)
        return cur

    def commit(self) -> None:
        """Commit the current transaction."""
        self._check_open()
        self._execute_simple("COMMIT")

    def rollback(self) -> None:
        """Roll back the current transaction."""
        self._check_open()
        self._execute_simple("ROLLBACK")

    def ping(self, timeout: float = 5.0) -> bool:
        """Send a Ping and wait for Pong."""
        try:
            self._check_open()
            self._send_frame(MSG_PING, b"")
            old_timeout = self._sock.gettimeout()  # type: ignore[union-attr]
            self._sock.settimeout(timeout)  # type: ignore[union-attr]
            try:
                msg_type, _ = self._recv_frame()
                return msg_type == MSG_PONG
            finally:
                self._sock.settimeout(old_timeout)  # type: ignore[union-attr]
        except Exception:
            return False

    def close(self) -> None:
        """Graceful disconnect."""
        if not self._closed and self._sock is not None:
            self._closed = True
            try:
                self._send_frame(MSG_DISCONNECT, b"")
                self._sock.settimeout(1.0)
                try:
                    self._recv_frame()
                except Exception:
                    pass
            finally:
                self._sock.close()
                self._sock = None

    def __enter__(self) -> Connection:
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()

    def __del__(self) -> None:
        if not self._closed:
            self.close()

    # ── Internal: query execution ────────────────────────────────────

    def _next_request_id(self) -> int:
        self._request_id += 1
        return self._request_id

    def _execute_query(
        self, sql: str, params: list[Any] | None = None
    ) -> tuple[list[ColumnMeta], list[list[Any]], int]:
        """Execute a query and return (columns, rows, rows_affected)."""
        self._check_open()
        req_id = self._next_request_id()
        session_flags = SESSION_AUTOCOMMIT if self._autocommit else 0

        w = WireWriter()
        w.u64(req_id)
        w.u64(self._server_epoch)
        w.string_u32(sql)
        # No binary params for now (simplified SDK)
        w.u16(0)
        w.u32(session_flags)

        self._send_frame(MSG_QUERY_REQUEST, w.to_bytes())
        return self._read_query_response()

    def _execute_simple(self, sql: str) -> int:
        """Execute a statement and return rows_affected."""
        _, _, affected = self._execute_query(sql)
        return affected

    def _read_query_response(
        self,
    ) -> tuple[list[ColumnMeta], list[list[Any]], int]:
        """Read a QueryResponse or ErrorResponse."""
        msg_type, payload = self._recv_frame()

        if msg_type == MSG_ERROR_RESPONSE:
            self._raise_error(payload)

        if msg_type != MSG_QUERY_RESPONSE:
            raise FalconError(
                f"unexpected response type: 0x{msg_type:02x}",
                error_code=0,
            )

        reader = WireReader(payload)
        _req_id = reader.u64()
        num_cols = reader.u16()

        columns: list[ColumnMeta] = []
        col_types: list[int] = []
        for _ in range(num_cols):
            name = reader.string_u16()
            type_id = reader.u8()
            nullable = reader.u8() != 0
            precision = reader.u16()
            scale = reader.u16()
            columns.append(ColumnMeta(name, type_id, nullable, precision, scale))
            col_types.append(type_id)

        num_rows = reader.u32()
        rows: list[list[Any]] = []
        for _ in range(num_rows):
            rows.append(decode_row(reader, col_types))

        rows_affected = reader.u64()
        return columns, rows, rows_affected

    # ── Internal: connection setup ───────────────────────────────────

    def _connect(self, timeout: float) -> None:
        """Establish TCP connection, handshake, and authenticate."""
        self._sock = socket.create_connection(
            (self._host, self._port), timeout=timeout
        )
        self._sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)

        self._handshake()

        if self._ssl_enabled and (self._negotiated_features & FEATURE_TLS):
            self._upgrade_tls()

        self._authenticate()

    def _handshake(self) -> None:
        w = WireWriter()
        w.u16(0)  # version major
        w.u16(1)  # version minor
        features = (
            FEATURE_BATCH_INGEST
            | FEATURE_PIPELINE
            | FEATURE_EPOCH_FENCING
            | FEATURE_BINARY_PARAMS
        )
        if self._ssl_enabled:
            features |= FEATURE_TLS
        w.u64(features)
        w.string_u16("falcondb-python/0.2")
        w.string_u16(self._database)
        w.string_u16(self._user)
        w.raw(os.urandom(16))  # nonce
        w.u16(0)  # num_params

        self._send_frame(MSG_CLIENT_HELLO, w.to_bytes())

        # Read ServerHello
        msg_type, payload = self._recv_frame()
        if msg_type == MSG_ERROR_RESPONSE:
            self._raise_error(payload)
        if msg_type != MSG_SERVER_HELLO:
            raise FalconError(f"expected ServerHello, got 0x{msg_type:02x}")

        reader = WireReader(payload)
        _server_major = reader.u16()
        _server_minor = reader.u16()
        self._negotiated_features = reader.u64()
        self._server_epoch = reader.u64()
        self._server_node_id = reader.u64()
        _server_nonce = reader.raw(16)
        num_params = reader.u16()
        for _ in range(num_params):
            reader.string_u16()  # key
            reader.string_u16()  # value

        # Read AuthRequest
        msg_type, payload = self._recv_frame()
        if msg_type != MSG_AUTH_REQUEST:
            raise FalconError(f"expected AuthRequest, got 0x{msg_type:02x}")
        reader = WireReader(payload)
        _auth_method = reader.u8()
        # Consume remaining challenge bytes
        if reader.remaining > 0:
            reader.raw(reader.remaining)

    def _authenticate(self) -> None:
        w = WireWriter()
        w.u8(AUTH_PASSWORD)
        w.raw(self._password.encode("utf-8"))
        self._send_frame(MSG_AUTH_RESPONSE, w.to_bytes())

        msg_type, payload = self._recv_frame()
        if msg_type == MSG_AUTH_OK:
            return
        if msg_type == MSG_AUTH_FAIL:
            reason = ""
            if payload:
                reader = WireReader(payload)
                reason = reader.string_u16()
            raise AuthenticationError(
                f"authentication failed: {reason}",
                error_code=4000,
                sqlstate="28P01",
            )
        raise FalconError(f"unexpected auth response: 0x{msg_type:02x}")

    def _upgrade_tls(self) -> None:
        ctx = ssl.create_default_context()
        if not self._ssl_verify:
            ctx.check_hostname = False
            ctx.verify_mode = ssl.CERT_NONE
        self._sock = ctx.wrap_socket(self._sock, server_hostname=self._host)  # type: ignore[arg-type]

    # ── Internal: framing ────────────────────────────────────────────

    def _send_frame(self, msg_type: int, payload: bytes) -> None:
        frame = build_frame(msg_type, payload)
        self._sock.sendall(frame)  # type: ignore[union-attr]

    def _recv_frame(self) -> tuple[int, bytes]:
        header = self._recv_exact(5)
        msg_type = header[0]
        length = struct.unpack_from("<I", header, 1)[0]
        if length > MAX_FRAME_SIZE:
            raise FalconError(f"frame too large: {length}")
        payload = self._recv_exact(length) if length > 0 else b""
        return msg_type, payload

    def _recv_exact(self, n: int) -> bytes:
        buf = bytearray()
        while len(buf) < n:
            chunk = self._sock.recv(n - len(buf))  # type: ignore[union-attr]
            if not chunk:
                raise FalconError("connection closed by server")
            buf.extend(chunk)
        return bytes(buf)

    def _raise_error(self, payload: bytes) -> None:
        reader = WireReader(payload)
        _req_id = reader.u64()
        error_code = reader.u32()
        sqlstate = reader.raw(5).decode("ascii")
        _retryable = reader.u8()
        server_epoch = reader.u64()
        message = reader.string_u16()
        raise make_error(message, error_code, sqlstate, server_epoch)

    def _check_open(self) -> None:
        if self._closed or self._sock is None:
            raise FalconError("connection is closed")

File: clients/falcondb-node/src/connection.js
'use strict';

const net = require('node:net');
const tls = require('node:tls');
const crypto = require('node:crypto');

const {
  MSG_CLIENT_HELLO, MSG_SERVER_HELLO, MSG_AUTH_REQUEST, MSG_AUTH_RESPONSE,
  MSG_AUTH_OK, MSG_AUTH_FAIL, MSG_QUERY_REQUEST, MSG_QUERY_RESPONSE,
  MSG_ERROR_RESPONSE, MSG_PING, MSG_PONG, MSG_DISCONNECT,
  FEATURE_BATCH_INGEST, FEATURE_PIPELINE, FEATURE_EPOCH_FENCING,
  FEATURE_TLS, FEATURE_BINARY_PARAMS,
  AUTH_PASSWORD, SESSION_AUTOCOMMIT,
  MAX_FRAME_SIZE, FRAME_HEADER_SIZE,
  WireWriter, WireReader, buildFrame, decodeRow,
} = require('./wire');

const { FalconError, AuthenticationError, makeError } = require('./errors');

/**
 * @typedef {Object} ConnectionConfig
 * @property {string}  [host='localhost']
 * @property {number}  [port=6543]
 * @property {string}  [database='falcon']
 * @property {string}  [user='falcon']
 * @property {string}  [password='']
 * @property {number}  [connectTimeout=10000]  - milliseconds
 * @property {boolean} [ssl=false]
 * @property {Object}  [sslOptions={}]         - options passed to tls.connect()
 * @property {boolean} [autocommit=true]
 */

/**
 * @typedef {Object} ColumnMeta
 * @property {string}  name
 * @property {number}  typeId
 * @property {boolean} nullable
 * @property {number}  precision
 * @property {number}  scale
 */

/**
 * @typedef {Object} QueryResult
 * @property {ColumnMeta[]} columns
 * @property {Array[]}      rows
 * @property {number}       rowsAffected
 */

class Connection {
  /**
   * @param {ConnectionConfig} config
   */
  constructor(config = {}) {
    this._host = config.host || 'localhost';
    this._port = config.port || 6543;
    this._database = config.database || 'falcon';
    this._user = config.user || 'falcon';
    this._password = config.password || '';
    this._connectTimeout = config.connectTimeout || 10000;
    this._ssl = config.ssl || false;
    this._sslOptions = config.sslOptions || {};
    this._autocommit = config.autocommit !== undefined ? config.autocommit : true;

    /** @type {net.Socket|tls.TLSSocket|null} */
    this._socket = null;
    this._closed = false;
    this._requestId = 0n;
    this._serverEpoch = 0n;
    this._serverNodeId = 0n;
    this._negotiatedFeatures = 0n;

    // Internal receive buffer — chunk list to avoid O(n²) Buffer.concat per data event
    this._recvChunks = [];
    this._recvLen = 0;
    this._recvBuf = Buffer.alloc(0);
    this._recvResolve = null;
    this._recvReject = null;
  }

  // ── Public API ─────────────────────────────────────────────────────────

  get autocommit() { return this._autocommit; }
  set autocommit(v) { this._autocommit = v; }
  get serverEpoch() { return this._serverEpoch; }
  get serverNodeId() { return this._serverNodeId; }
  get closed() { return this._closed; }

  /**
   * Connect to the FalconDB server.
   * @returns {Promise<void>}
   */
  async connect() {
    await this._tcpConnect();

    // Upgrade to TLS BEFORE sending any application data (handshake contains
    // database name, user name, nonce) so credentials are never sent in plaintext.
    if (this._ssl) {
      await this._upgradeTls();
    }

    this._setupRecv();
    await this._handshake();
    await this._authenticate();
  }

  /**
   * Execute a SQL query and return the result.
   * @param {string} sql
   * @returns {Promise<QueryResult>}
   */
  async query(sql) {
    this._checkOpen();
    const reqId = this._nextRequestId();
    const sessionFlags = this._autocommit ? SESSION_AUTOCOMMIT : 0;

    const w = new WireWriter();
    w.u64(reqId);
    w.u64(this._serverEpoch);
    w.stringU32(sql);
    w.u16(0); // no params
    w.u32(sessionFlags);

    this._sendFrame(MSG_QUERY_REQUEST, w.toBuffer());
    return this._readQueryResponse();
  }

  /**
   * Execute a SQL statement and return the number of rows affected.
   * @param {string} sql
   * @returns {Promise<number>}
   */
  async exec(sql) {
    const result = await this.query(sql);
    return result.rowsAffected;
  }

  /**
   * Commit the current transaction.
   * @returns {Promise<void>}
   */
  async commit() {
    await this.exec('COMMIT');
  }

  /**
   * Roll back the current transaction.
   * @returns {Promise<void>}
   */
  async rollback() {
    await this.exec('ROLLBACK');
  }

  /**
   * Send a Ping and wait for Pong.
   * @param {number} [timeout=5000] - milliseconds
   * @returns {Promise<boolean>}
   */
  async ping(timeout = 5000) {
    try {
      this._checkOpen();
      this._sendFrame(MSG_PING, Buffer.alloc(0));
      const { msgType } = await this._recvFrameWithTimeout(timeout);
      return msgType === MSG_PONG;
    } catch {
      return false;
    }
  }

  /**
   * Gracefully close the connection.
   * @returns {Promise<void>}
   */
  async close() {
    if (this._closed || !this._socket) return;
    this._closed = true;
    try {
      this._sendFrame(MSG_DISCONNECT, Buffer.alloc(0));
      await this._recvFrameWithTimeout(1000).catch(() => {});
    } finally {
      this._socket.destroy();
      this._socket = null;
    }
  }

  // ── Internal: TCP connection ───────────────────────────────────────────

  /** @returns {Promise<void>} */
  _tcpConnect() {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new FalconError(`connect timeout after ${this._connectTimeout}ms`));
        if (this._socket) this._socket.destroy();
      }, this._connectTimeout);

      this._socket = net.createConnection(
        { host: this._host, port: this._port, noDelay: true },
        () => {
          clearTimeout(timer);
          resolve();
        },
      );

      this._socket.once('error', (err) => {
        clearTimeout(timer);
        reject(new FalconError(`connect error: ${err.message}`));
      });
    });
  }

  _setupRecv() {
    this._socket.on('data', (chunk) => {
      // Accumulate chunks in a list — O(1) per data event instead of O(n) concat
      this._recvChunks.push(chunk);
      this._recvLen += chunk.length;
      this._tryDeliver();
    });

    this._socket.on('error', (err) => {
      if (this._recvReject) {
        this._recvReject(new FalconError(`socket error: ${err.message}`));
        this._recvReject = null;
        this._recvResolve = null;
      }
    });

    this._socket.on('close', () => {
      if (this._recvReject) {
        this._recvReject(new FalconError('connection closed by server'));
        this._recvReject = null;
        this._recvResolve = null;
      }
    });
  }

  /** Consolidate chunk list into a single contiguous buffer when needed. */
  _consolidateRecv() {
    if (this._recvChunks.length === 0) return;
    if (this._recvBuf.length === 0) {
      this._recvBuf = this._recvChunks.length === 1
        ? this._recvChunks[0]
        : Buffer.concat(this._recvChunks, this._recvLen);
    } else {
      this._recvBuf = Buffer.concat(
        [this._recvBuf, ...this._recvChunks],
        this._recvBuf.length + this._recvLen,
      );
    }
    this._recvChunks = [];
    this._recvLen = 0;
  }

  _tryDeliver() {
    if (!this._recvResolve) return;

    // Only consolidate when we actually need to parse
    this._consolidateRecv();

    if (this._recvBuf.length < FRAME_HEADER_SIZE) return;

    const msgType = this._recvBuf.readUInt8(0);
    const length = this._recvBuf.readUInt32LE(1);

    if (length > MAX_FRAME_SIZE) {
      const reject = this._recvReject;
      this._recvResolve = null;
      this._recvReject = null;
      reject(new FalconError(`frame too large: ${length}`));
      return;
    }

    if (this._recvBuf.length < FRAME_HEADER_SIZE + length) return;

    const payload = this._recvBuf.subarray(FRAME_HEADER_SIZE, FRAME_HEADER_SIZE + length);
    this._recvBuf = this._recvBuf.subarray(FRAME_HEADER_SIZE + length);

    const resolve = this._recvResolve;
    this._recvResolve = null;
    this._recvReject = null;
    resolve({ msgType, payload: Buffer.from(payload) });
  }

  /**
   * @returns {Promise<{msgType: number, payload: Buffer}>}
   */
  _recvFrame() {
    return new Promise((resolve, reject) => {
      this._recvResolve = resolve;
      this._recvReject = reject;
      // Check if data already buffered
      this._tryDeliver();
    });
  }

  /**
   * @param {number} ms
   * @returns {Promise<{msgType: number, payload: Buffer}>}
   */
  _recvFrameWithTimeout(ms) {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this._recvResolve = null;
        this._recvReject = null;
        reject(new FalconError(`receive timeout after ${ms}ms`));
      }, ms);

      this._recvResolve = (result) => {
        clearTimeout(timer);
        resolve(result);
      };
      this._recvReject = (err) => {
        clearTimeout(timer);
        reject(err);
      };
      this._tryDeliver();
    });
  }

  /**
   * @param {number} msgType
   * @param {Buffer} payload
   */
  _sendFrame(msgType, payload) {
    const frame = buildFrame(msgType, payload);
    this._socket.write(frame);
  }

  // ── Internal: handshake ────────────────────────────────────────────────

  async _handshake() {
    const w = new WireWriter();
    w.u16(0); // version major
    w.u16(1); // version minor
    let features = FEATURE_BATCH_INGEST | FEATURE_PIPELINE |
                   FEATURE_EPOCH_FENCING | FEATURE_BINARY_PARAMS;
    if (this._ssl) features |= FEATURE_TLS;
    w.u64(features);
    w.stringU16('falcondb-node/0.2');
    w.stringU16(this._database);
    w.stringU16(this._user);
    w.raw(crypto.randomBytes(16)); // nonce
    w.u16(0); // num_params

    this._sendFrame(MSG_CLIENT_HELLO, w.toBuffer());

    // Read ServerHello
    const hello = await this._recvFrame();
    if (hello.msgType === MSG_ERROR_RESPONSE) {
      this._raiseError(hello.payload);
    }
    if (hello.msgType !== MSG_SERVER_HELLO) {
      throw new FalconError(`expected ServerHello, got 0x${hello.msgType.toString(16)}`);
    }

    const r = new WireReader(hello.payload);
    r.u16(); // server major
    r.u16(); // server minor
    this._negotiatedFeatures = r.u64();
    this._serverEpoch = r.u64();
    this._serverNodeId = r.u64();
    r.raw(16); // server nonce
    const numParams = r.u16();
    for (let i = 0; i < numParams; i++) {
      r.stringU16(); // key
      r.stringU16(); // value
    }

    // Read AuthRequest
    const authReq = await this._recvFrame();
    if (authReq.msgType !== MSG_AUTH_REQUEST) {
      throw new FalconError(`expected AuthRequest, got 0x${authReq.msgType.toString(16)}`);
    }
    // Consume auth method + challenge
    const ar = new WireReader(authReq.payload);
    ar.u8(); // auth method
    if (ar.remaining > 0) ar.raw(ar.remaining);
  }

  async _authenticate() {
    const w = new WireWriter();
    w.u8(AUTH_PASSWORD);
    w.raw(Buffer.from(this._password, 'utf-8'));
    this._sendFrame(MSG_AUTH_RESPONSE, w.toBuffer());

    const resp = await this._recvFrame();
    if (resp.msgType === MSG_AUTH_OK) return;
    if (resp.msgType === MSG_AUTH_FAIL) {
      let reason = '';
      if (resp.payload.length > 0) {
        const r = new WireReader(resp.payload);
        reason = r.stringU16();
      }
      throw new AuthenticationError(
        `authentication failed: ${reason}`, 4000, '28P01', 0n,
      );
    }
    throw new FalconError(`unexpected auth response: 0x${resp.msgType.toString(16)}`);
  }

  async _upgradeTls() {
    return new Promise((resolve, reject) => {
      const opts = {
        ...this._sslOptions,
        socket: this._socket,
        servername: this._sslOptions.servername || this._host,
      };
      const tlsSocket = tls.connect(opts, () => {
        this._socket = tlsSocket;
        this._setupRecv();
        resolve();
      });
      tlsSocket.once('error', reject);
    });
  }

  // ── Internal: query execution ──────────────────────────────────────────

  /**
   * @returns {Promise<QueryResult>}
   */
  async _readQueryResponse() {
    const { msgType, payload } = await this._recvFrame();

    if (msgType === MSG_ERROR_RESPONSE) {
      this._raiseError(payload);
    }

    if (msgType !== MSG_QUERY_RESPONSE) {
      throw new FalconError(`unexpected response type: 0x${msgType.toString(16)}`);
    }

    const reader = new WireReader(payload);
    reader.u64(); // request_id
    const numCols = reader.u16();

    /** @type {ColumnMeta[]} */
    const columns = [];
    const colTypes = [];
    for (let i = 0; i < numCols; i++) {
      const name = reader.stringU16();
      const typeId = reader.u8();
      const nullable = reader.u8() !== 0;
      const precision = reader.u16();
      const scale = reader.u16();
      columns.push({ name, typeId, nullable, precision, scale });
      colTypes.push(typeId);
    }

    const numRows = reader.u32();
    const rows = [];
    for (let i = 0; i < numRows; i++) {
      rows.push(decodeRow(reader, colTypes));
    }

    const rowsAffected = Number(reader.u64());
    return { columns, rows, rowsAffected };
  }

  /**
   * @param {Buffer} payload
   */
  _raiseError(payload) {
    // Minimum error frame: reqId(8) + code(4) + sqlstate(5) + retryable(1) + epoch(8) + msgLen(2) = 28
    if (payload.length < 28) {
      throw new FalconError(`malformed error response (${payload.length} bytes, need >= 28)`);
    }
    const r = new WireReader(payload);
    r.u64(); // request_id
    const errorCode = r.u32();
    const sqlstate = r.raw(5).toString('ascii');
    r.u8(); // retryable
    const serverEpoch = r.u64();
    const message = r.stringU16();
    throw makeError(message, errorCode, sqlstate, serverEpoch);
  }

  _nextRequestId() {
    this._requestId += 1n;
    return this._requestId;
  }

  _checkOpen() {
    if (this._closed || !this._socket) {
      throw new FalconError('connection is closed');
    }
  }
}

/**
 * Create and connect to a FalconDB server.
 * @param {ConnectionConfig} config
 * @returns {Promise<Connection>}
 */
async function connect(config = {}) {
  const conn = new Connection(config);
  await conn.connect();
  return conn;
}

module.exports = { Connection, connect };
